python/cudf/cudf/core/buffer/buffer.py

```diff
@@ -171,6 +172,37 @@ def __getitem__(self, key: slice) -> Buffer:
             raise ValueError("slice must be C-contiguous")
         return self._getitem(offset=start, size=stop - start)
 
+    def copy(self, deep: bool = True):
+        """
+        Return a copy of Buffer.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            If True, returns a deep-copy of the underlying Buffer data.
+            If False, returns a shallow-copy of the Buffer pointing to
+            the same underlying data.
+
+        Returns
+        -------
+        Buffer
+        """
+        if not deep:
+            shallow_copy = Buffer.__new__(Buffer)
+            shallow_copy._ptr = self._ptr
+            shallow_copy._size = self._size
+            shallow_copy._owner = self._owner
+            return shallow_copy
+        else:
+            owner_copy: rmm.DeviceBuffer = copy.copy(self._owner)
```

Comment:
We cannot always assume that `copy.copy()` creates a copy of the device data underlying `self._owner`. Yes, `DeviceBuffers` and `cupy.ndarray` objects have the behavior that `copy.copy()` creates a deep copy of them, but, for example, Numba device arrays do not:

```python
In [1]: from numba import cuda

In [2]: import copy

In [3]: a = cuda.to_device([1, 2, 3])

In [4]: a.__cuda_array_interface__
Out[4]: 
{'shape': (3,),
 'strides': None,
 'data': (140652119588864, False),
 'typestr': '<i8',
 'stream': None,
 'version': 3}

In [5]: copy.copy(a).__cuda_array_interface__
Out[5]: 
{'shape': (3,),
 'strides': None,
 'data': (140652119588864, False),
 'typestr': '<i8',
 'stream': None,
 'version': 3}
```

Reply:
Should we special case this behavior for Numba device arrays? or I think a better approach would be to convert all owner objects here to a `DeviceBuffer` and then perform a copy?

Reply:
Yikes, just read @madsbk comment below which is the same :) So applying that change.